02-Context-or-Narrative/Good-Documentation.md

```diff
@@ -0,0 +1,135 @@
+
+Did someone ever teach you how to write "good" documentation? Probably not - you likely just wrote what came to mind and hoped it was good enough.
+
+Because of this, it's useful to learn a definition of good documentation and understand why it's defined that way.
+
+Essentially, there are 3 factors that affect whether documentation is "good" or not.
+1. Its intended audience
+2. Its type: the question being answered, targeting a specific subsection of the audience
+3. How accurately it reflects the desired version of the code/project
+
+It's important to note that this section is sourced from only two blog posts, [What Nobody Tells You About Documentation](https://www.divio.com/blog/documentation/) and [Teach, Don't Tell](http://stevelosh.com/blog/2013/09/teach-dont-tell/). As we learn more this section might change in significant ways, so any conclusions we draw should be kept "flexible".
+
+## The Types of Documentation
+
+First, there are 5 types of documentation that target specific phases of a learner's experience (as explained in [What Nobody Tells You About Documentation](https://www.divio.com/blog/documentation/) and [Teach, Don't Tell](http://stevelosh.com/blog/2013/09/teach-dont-tell/))
+
+| Learner's Phase | Type | Analogy | Characteristics
+| -- | -- | -- | -- |
+| Curious Outsider | The Hook | Selling a product to a potential customer | Answers these questions: <ul><li>What is this thing? / What problem does it solve?</li><li>Why whould I care? / How is this relevant to me for my purposes? / Who should not care?</li><li>How long will it take to learn it and how difficult is the learning curve?</li><li>Where do I go to get started / learn how to use this?</li></ul>
+| Potential User | Getting Started | Teaching a child how to cook | <ul><li>Focuses on the learner 'doing' stuff, not 'explaining' stuff to the learner</li><li>Provides a small simple working example that teaches the basics</li><li>New learners experience an 'I can use this now!' moment by the end</li><li>Focuses on concrete tasks, not abstract concepts</li><li>Does not use jargon</li><li>Explains only what is necessary and cuts out all else</li><li>Avoids explaining deeper concepts or different ways of doing the same thing</li></ul>
+| New User | How-To Guides | Following a cookbook's recipe | <ul><li>Achieves some goal or solves a problem</li><li>States the pre-requisites one needs to have before starting (not a Getting Started Guide)</li><li>The Guide follows a clearly-labeled step-by-step process</li><li>By following the steps, one reproduces the same results without fail</li><li>Explains the different ways one can achieve the same goal</li><li>Explains only what is necessary</li></ul>
+| Active User | Reference | Reading an encyclopedia | <ul><li>Concise explanation of each piece of the code</li><li>The structure of the reference mirrors the structure of the code it documents</li><li>Formatting is consistent throughout the material</li></ul>
```

Comment:
I feel like there's another important type of documentation for "Active User" than just "Reference". For example, I'm often looking for ideas of how to solve certain problems. Perhaps this type of documentation doesn't belong in language-level documentation, but "up" a level in a group-level documentation. For example, if I'm trying to write web-form input validation and I don't know the various ways of doing that, I'd like to ask others who have done that before. That "others" would form a group, and that group would need to include an answer to that question in its documentation.

So, would it be good to reference "groups" here? I'll think about it and come back later with a follow-up comment and concrete suggestion if it's still a good idea.

Reply:
> I'm often looking for ideas of how to solve certain problems. 

I'd label that as a 'Explanation' doc type. 

Also, interpreting that table in a linear fashion is misleading. The 'what nobody tells you about documentation' source actually grouped these together using an XY graph where one axis was theory/practical axis and the other was the 'when we're studying / when we're working' axis

Perhaps the issue here is the 'Experienced User' label we've used to identify the person that corresponds to the Explanation doc type. That might make it seem like one needs to have gone through all the how-to guides fully and have used the library for a while before they are 'qualified' to ask such questions.

Reply:
It sounds like you also feel this table could be clarified in some way. I'll resolve this thread and remind myself to come back to it sometime in the future with this issue: https://github.com/chexxor/purescript-documentation-discussion/issues/70